src/api-builder.ts

```
@@ -0,0 +1,182 @@
+        return this.pendingRequests < queueLimit;
+    }
+
+    protected clearQueue(): void {
+        this.requestsQueue = [];
+    }
+
+    protected onIdentityLogout(): void {
+        this.clearQueue();
+    }
+
+    // tslint:disable-next-line:no-any
+    private urlBuilder(request: ApiRequest<any>): string {
+        const path: string = this.configuration.path != null ? this.configuration.path : "";
+        let requestPath: string = "";
+        let requestQueryString: string = "?";
+
+        if (request.requestPath != null) {
+            requestPath = request.requestPath;
+        }
+
+        if (this.configuration.defaultQueryParams != null) {
+            requestQueryString += queryString.stringify(this.configuration.defaultQueryParams);
+        }
+
+        if (request.queryParams != null) {
+            requestQueryString += queryString.stringify(request.queryParams);
+        }
+
+        if (requestQueryString === "?") {
+            requestQueryString = "";
+        }
+
+        return `${this.configuration.host}${path}${requestPath}${requestQueryString}`;
+    }
+
+    protected async makeRequest(): Promise<void> {
+        let request: QueuedRequest;
+
+        const forceRequestIndex = this.requestsQueue.findIndex(x => x.isForced === true);
+        const canMakeRequest = this.canMakeRequest();
+
+        // If there are forced requests waiting in the queue.
+        if (forceRequestIndex !== -1) {
+            // Perform them first no matter whether we're allowed to make requests.
+            // Take force request out of the queue.
+            request = this.requestsQueue.splice(forceRequestIndex, 1)[0];
+        } else if (canMakeRequest) {
+            // Simply take FIFO request.
+            const nextInQueue = this.requestsQueue.shift();
+            if (nextInQueue == null) {
+                return;
+            }
+            request = nextInQueue;
+        } else {
+            return;

```

Remake this `if` to fail fast.